### ISSUE TITLE
Error in test_hmc.py?

### ISSUE THREAD

**Issue by junpenglao** [uber/pyro#697]
Hi there, I am playing around with the recently merged HMC and I think there might be a mistake in the test:
https://github.com/uber/pyro/blob/7b3a0a833b836c1b7b1e981091b514baec819c40/tests/infer/mcmc/test_hmc.py#L35-L37

Judging from the trace, seems only the node `mu_4` in the graph was being sampled correctly, the other nodes `mu_1` ~ `mu_3` is not in the graph:

I also put up a gist that demonstrates this:
https://gist.github.com/junpenglao/013255dee83f39d9e05f72b9b0aa81e9#file-HMC_in_pyro-ipynb

**Comment by neerajprad** [comment]
Thanks for trying out HMC! The tests seem to be generating the correct samples for all the latents. A couple of points for your particular example:

 - HMC is very sensitive to the `step_size` and `num_steps` params, and you may need to adjust it for your example. Looking at the plots, I do not think the values are configured correctly. See the gist below for the parameter values that I used.
 - I ran the example that you linked above. You can already start seeing the separation of means along the different dimensions by the time you reach `mu_2` (`mu_3` is even clearer). The difference is that the variance of the intermediate latents is much higher than `mu_4` and you will observe some noise. This should be expected as `mu_4` is directly connected to the observations. 

I have not worked with PyMC3, but do let us know if you observe any inconsistencies.

https://gist.github.com/neerajprad/f12514cefc4f0601aabf99b839d2f823

**Comment by junpenglao** [comment]
Thanks for the information @neerajprad! And great work implementing HMC in PyTorch ;-)

- I agree that HMC is very sensitive to hyperparameters, which is why I am actually using the parameters tuned from `PyMC3` to initiate the sampler (it would work even better if the HMC also tune the mass matrix). Unless the parameters mean different things in pyro and pymc3, the tuned parameters should work well.

- I think this is my confusion: why is only `mu_4` directly connected to the observations? If you are sampling from the posterior of `mu`, it should be connected to the observation otherwise you are just sampling from the prior. As you can see in my gist from cell 19, connecting `mu` to the observation seems to work as all trace now gives the same posterior.

Another observation I find a bit puzzling is that there is these spindle shapes on the trace, which normally should be there...

**Comment by neerajprad** [comment]
> I agree that HMC is very sensitive to hyperparameters, which is why I am actually using the parameters tuned from PyMC3 to initiate the sampler (it would work even better if the HMC also accept the tuned mass matrix). Unless the parameters mean different things in pyro and pymc3, the tuned parameters should work well.

I will need to study the PyMC3 implementation in detail to see the differences, but I would not expect the parameters to work directly, as there could be a lot of subtle differences in the implementation. As an example of how sensitive this implementation is to the choice of `step_size`, changing this param in the gist that I posted above from `0.165` to `0.15` or `0.18` will give results that are completely off. You are also right that the mass matrix is also assumed to be identity in this base HMC kernel. 

> I think this is my confusion: why is only mu_4 directly connected to the observations? If you are sampling from the posterior of mu, it should be connected to the observation otherwise you are just sampling from the prior. 

I think there might be some confusion about the model. The graphical model is as follows: 
```
0    1
|   /
mu_1   1
|     /
|   /
mu_2   1
|    /
|  /
mu_3   1
|    /
|  /
mu_4   1
|    /
|  /
[observations]

mu_i ~ N(mu_{i-1}, 1)
```
As such, `mu_4` is the one that is directly connected to the observations. The posterior of the remaining latents will certainly take into account the observed values (which they do as you can see from the last figure in the gist that I posted above, and the more obvious updated one below). However, the posterior variance for `mu_4` will be lower than the remaining ones as you increase the number of observations (will go to 0 in the limit). To see this, you could try out a simpler example having the same structure, where you have only two intermediate latents - `mu_1` and `mu_2`, and observe their posterior variance as you change the number of observations from 1 to 1000. You should observe the same results in PyMC3 as well. I have modified your example in this gist (https://gist.github.com/neerajprad/5e431a7eb24186f92cee9acd6e858381) by scaling the observed values by 10, to make it obvious what is happening. You will clearly be able to see the values sampled from the intermediate latents and how the variance changes, in the last figure.

> As you can see in my gist from cell 19, connecting mu to the observation seems to work as all trace now gives the same posterior.

Cell 19 does not seem to be the same chained graphical model?

> Another observation I find a bit puzzling is that there is these spindle shapes on the trace

The shape for the posterior trace samples is as you would expect - the latents are interpolating between the prior (all set to 0) and the posterior (arange(dim)). Does the figure from the updated gist make it clearer?

**Comment by junpenglao** [comment]
Oh I see! I confused the term `chain`, as in `PyMC3` we use `chain` to indicate different parallel MCMC chain to sample. But here it is a chain of node which mu_0 --> mu_1 --> mu_2 --> mu_3 --> mu_4 --> obs right?

**Comment by neerajprad** [comment]
> But here it is a chain of node which mu_0 --> mu_1 --> mu_2 --> mu_3 --> mu_4 --> obs right?

That's right. :) I should have mentioned earlier. Stan uses the term `chain` to refer to parallel MCMC runs too. CPU parallelization is something we haven't explored yet.
